### PR TITLE
[TESTERS NEEDED] cellMic: wait for registration before using secondary microphones

### PR DESCRIPTION
**Pull request by Megamouse**
Apparently games have to register the second and subsequent microphones before they can use them.

### CODE REVIEW

**Comment by T3HM3T4L** [comment]
It took me a bit longer to test because I wasn't sure if I had built it correctly.

When I boot the game with a microphone selected in the GUI, RPCS3 crashes with
"F {PPU[0x1000018] Thread (ThePs3MicManager::RegisterMi) [HLE:0x011d02fc, LR:0x0087ca48]} SIG: Thread terminated due to fatal error: Verification failed
(in file /home/mooty/rpcs3_git/rpcs3/rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp:144, in function copy_device_info)"

If I try to boot the game with no microphone selected in the GUI, the game will boot normally.

**Comment by Megamouse** [comment]
@T3HM3T4L try again please

**Comment by T3HM3T4L** [comment]
> @T3HM3T4L try again please (you can just download the build, you don't have to build it yourself)

I'm a total brainlet so I haven't figured this out yet. Logically there had to be an easier way but I didn't find it.

**Comment by T3HM3T4L** [comment]
Could you please explain how I would go about just downloading the build? I still can't manage it.
I tried downloading the audio2 branch zip but I don't think that's it either.

**Comment by Darkhost1999** [comment]
> Could you please explain how I would go about just downloading the build?

https://wiki.rpcs3.net/index.php?title=Help:Using_different_versions_of_RPCS3#How_to_download_builds_from_a_pull_request_on_GitHub?

**Comment by T3HM3T4L** [comment]
> > Could you please explain how I would go about just downloading the build?
> 
> https://wiki.rpcs3.net/index.php?title=Help:Using_different_versions_of_RPCS3#How_to_download_builds_from_a_pull_request_on_GitHub?

Thanks a bunch!
Now it was easy-peasy.

Great news: seems to absolutely work!
I used a Singstar mic combo (one plugged in through Singstar's usb, the other via front mic input) and the game correctly picks them up as mic1 and mic2.
Booted up a song and it works as it should!

Currently testing with 3 mic inputs.

**Comment by YuriNator557** [comment]
Which Singstar game was this with? There are a lot and some are already known to work even before this PR

**Comment by T3HM3T4L** [comment]
> Which Singstar game was this with? There are a lot and some are already known to work even before this PR

This was Rock Band 3 actually. I know there's a Singstar and Singstar Real mode in the audio settings for those games, but these didn't work with Rock Band 3 at all.

2 mics seems to work very well and giving me the usual performance but although 3 mics works, there seems to be a huge performance hit (I'm running a Ryzen 7 3800x) from the moment the game boots to main screen and intermittently while the song is playing.

**Comment by T3HM3T4L** [comment]
Here are the facts I can line up:

2 mics in Rock Band 3: works as intended, no hit to performance (steady 30 FPS in menu, steady 60 FPS during song)
3 mics in Rock Band 3: does work as intended, but performance gets hit hard with intervals in the main menu (crashes down to as low 5 FPS, then recovers, audio stutters alongside these drops) and during song (even when I leave the game on pause, the FPS dips as low as 17 FPS at times, about 80% of the time it's at 60 FPS though).
This is with my CPU in performance mode and everything.

My total CPU utilization will be at 10-20% normally and whenever the FPS tanked the CPU utilization went up to anywhere between 20-40% total usage.

The only other thing that I can note is that the log usually spams
"U cellAvconfExt TODO: cellAudioInGetAvailableDeviceInfo(count=8, info=*0xd00bb830)
U cellAudio TODO: cellAudioSendAck(data3=0x0) x191"
during Rock Band 3, which I think is the game constantly pinging for new devices being hotplugged, which will very consistently have around x190 attempts per log entry line.
Any time the FPS dropped in the main menu, the attempts would drop to anywhere between x99-x170.

I'm trying to include a log but GitHub won't allow me as the text file is too big and it won't accept the log.gz.

There's a lot of
"{RSX [0x03b8cf8]} RSX: [Performance warning] CPU readback of swizzled data"
and
"RSX: Cache miss at address 0xC923AE00. This is gonna hurt..."

**Comment by Megamouse** [comment]
Thanks. Please focus on the mic stuff. The RSX isn't relevant.
This also needs testing with all the other mic input games whether they already worked or not.
There seem to be multiple issues for different games already

**Comment by T3HM3T4L** [comment]
> Thanks. Please focus on the mic stuff. The RSX isn't relevant. This also needs testing with all the other mic input games whether they already worked or not. There seem to be multiple issues for different games already

After retrying by closing RPCS3 sooner, I got a smaller log which I'm able to link here. The previous log had ballooned to 125MB (5 times the file size limit!) after about 30 seconds in the Rock Band 3 main menu, so something was spamming REALLY hard, which would explain the CPU having difficulties.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/10892664/RPCS3.log)

**Comment by Megamouse** [comment]
@T3HM3T4L I added something that will reduce the log spam.

**Comment by T3HM3T4L** [comment]
> @T3HM3T4L I added something that will reduce the log spam.

Here's the newest log from the newest build.
Still very strained 3 mic performance with a whole lot of "this is gonna hurt..." up to me closing the game.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/10899788/RPCS3.log)

File: rpcs3/Emu/Cell/Modules/cellAudio.cpp [hunk 1]
```diff
@@ -615,7 +616,7 @@ void cell_audio_thread::advance(u64 timestamp)
 	for (u32 i = 0; i < queue_count; i++)
 	{
 		lv2_obj::notify_all_t notify;
-		queues[i]->send(event_sources[i], 0, 0, 0);
+		queues[i]->send(event_sources[i], CELL_AUDIO_EVENT_MIX, 0, event_data3[i]);
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellAudio.cpp:619]: iirc when I tested on realhw these values were indeed 0. Iirc even documentation suggests they are unused and should not be read.

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellAudio.cpp:619]: CELL_AUDIO_EVENT_MIX is 0.
Data 3 is also 0 unless you use the before mix flag. It is used to acknowledge the audio mix for that queue

File: rpcs3/Emu/Cell/Modules/cellMic.cpp [hunk 2]
```diff
@@ -134,45 +133,77 @@ void mic_context::operator()()
...
+	case microphone_handler::singstar:
+	case microphone_handler::null:
+	default:
+		break;
+	}
+
+	return ::narrow<u32>(index);
+}
+
+void mic_context::unregister_device(u32 dev_num)
```
Comment (RipleyTom) [inline rpcs3/Emu/Cell/Modules/cellMic.cpp:194]: Hmm so let's say a game registers 2 extra mics.
We'd have mic 0, 1, 2. Then if it unregisters the mic with device 1 the device 2 mic becomes the device 1 mic?
I think that's probably incorrect.

**Comment by qfoxb** [comment]
Just tested this out with Rock Band 3, 3 mic input is working great for me now, performance seems the same
Also, the log spam fix reduced my log file size tenfold. 
The only mic related issue with Rock Band that i still have is the microphone audio dropping out randomly while singing long notes.
Will this PR close issue #8209?
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/10951549/RPCS3.log)

**Comment by DarkRTA** [comment]
Mic input works great on rock band 3 for me now, although the performance is still terrible.

**Comment by T3HM3T4L** [comment]
Tested the latest, 3 mics still hits the performance hard. The main menu should be at steady 30 FPS with 20% total CPU but the framerate will randomly tank into the 10's with total CPU going up to 30-40%.
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/10993200/RPCS3.log)

**Comment by rmf757** [comment]
Tested with 2 and 3 mics. On my pc (3060 + Ryzen 7 5800HS) it worked fine. I didn’t notice any appreciable performance losses when using all 3 mics either! Great work!

**Comment by T3HM3T4L** [comment]
> Tested with 2 and 3 mics. On my pc (3060 + Ryzen 7 5800HS) it worked fine. I didn’t notice any appreciable performance losses when using all 3 mics either! Great work!

Reports like these make me suspect it's going to be a Linux issue (realistic assumption I'm the only one running Linux here).

Here's a log on the latest main branch RPCS3 with all my packages updated, 3 mics active, trying to make the circumstances as similar to my audio2 branch testing.
I tried to let a full song play but I had to quit prematurely before the log ballooned over 25mb.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11010527/RPCS3.log)

**Comment by T3HM3T4L** [comment]
And here's the second log: audio2 branch, 3 mics active, played the same song as the main branch test.

main branch - 30 FPS steady main menu, 60 FPS steady during the song (up until the 25mb log cutoff), no audio crackling
audio2 branch - hard FPS drop in main menu with audio crackling, heavy slowdowns during the song with audio crackling, sometimes moving back up to 60 FPS for a bit and then more slowdowns

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11010538/RPCS3.log)

**Comment by Readek** [comment]
Just had a local party with this build, the game could not handle 3 mics so we played with 2, everything seemed fine aside from rare slowdowns mid-song that usually don't happen with default build.

On a side note, the game froze/crashed consistently after a certain time (had 5 or 6 freezes for the duration of the session) when loading songs. Unsure if this is because of this build, it could be unrelated because of a variety of other reasons (playing with RB3DX, playing with a higher than intended sound count, etc.)

**Comment by AARP41298** [comment]
Hi, i just do a one song test with the audio2 branch in RB3, and it works fantastic

I hope to make some time over the weekend to do some more testing with RB3DX, customs and the beatles.

**Comment by jusman010** [comment]
Had some friends over yesterday. Worked all night with 3 mics - no lag. RB3DX and customs included.

**Comment by qfoxb** [comment]
did you use official rock band microphones or third party ones? just curious

File: rpcs3/Emu/Cell/Modules/cellMic.cpp (hunk 2 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellMic.cpp:213]: Changed to use a fixed amount of mics.
We'll see if this works...

**Comment by Megamouse** [comment]
Thanks for testing so extensively.
I changed something, so now it needs to be tested again after the builds are finished :)

**Comment by Readek** [comment]
Tested the new build, but it did sadly not fix the performance issues (unplayable with 3 mics, rare slowdowns with 2 mics).

I forgot to add that that last test (as well as this one) was done on Linux.

**Comment by Megamouse** [comment]
The performance is irrelevant in this PR

**Comment by T3HM3T4L** [comment]
> Tested the new build, but it did sadly not fix the performance issues (unplayable with 3 mics, rare slowdowns with 2 mics).
> 
> I forgot to add that that last test (as well as this one) was done on Linux.

These were exactly my findings on Linux as well as earlier reported in the thread.
New build indeed doesn't seem to change anything in that regard.
So far, it's looking like 100% performance issue rate on Linux and 0% on Windows, as far as I know.

I also tested with 1 mic which gave rocksteady performance on par with the main branch so if the only problem is performance issues on Linux specifically while using the optional 2 or 3 mic setting, this PR is still an easy net positive for Windows users and a status quo for Linux users.

File: rpcs3/Emu/Cell/Modules/cellMic.cpp [hunk 3]
```diff
@@ -134,48 +134,109 @@ void mic_context::operator()()
...
-			::at32(mic_list, 0).add_device(device_list[0]);
+			index = i;
 			break;
 		}
-		case microphone_handler::null:
-		default: break;
+	}
+
+	// TODO: Check max mics properly
+	ensure(index <= mic_list.size(), "cellMic max mics exceeded during registration");
```
Comment (RipleyTom) [inline rpcs3/Emu/Cell/Modules/cellMic.cpp:184]: It should be < mic_list.size() also with current logic if no free spot is found index will be 0 which is not correct.

File: rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp [hunk 4]
```diff
@@ -368,12 +395,31 @@ error_code cellAudioInRegisterDevice(u64 deviceType, vm::cptr<char> name, vm::pt
...
+	std::optional<CellAudioInDeviceInfo> info = av_manager.get_device_info(name);
+	if (!info || !memchr(info->name, '\0', sizeof(info->name)))
+	{
+		// TODO
+		return CELL_AUDIO_IN_ERROR_DEVICE_NOT_FOUND;
+	}
+
+	auto& mic_thr = g_fxo->get<mic_thread>();
+	const std::lock_guard mic_lock(mic_thr.mutex);
+	const u32 device_number = mic_thr.register_device(info->name);
```
Comment (RipleyTom) [inline rpcs3/Emu/Cell/Modules/cellAvconfExt.cpp:410]: From doc I think the game can repeatedly call this function, it'll just return the same device number again.
We're missing a check to make sure not to register the same device twice, either here or in mic_context::register_device.

**Comment by qfoxb** [comment]
Just tested this build with one mic with Rock Band 3. cellAudioInGetAvailableDeviceInfo is still printed a lot in the log. Not sure if it's in the scope of this PR but changing the microphone settings in custom configuration while the game is open will not actually change the microphones in game. 
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11069126/RPCS3.log)

**Comment by Megamouse** [comment]
Needs another test. Fixed some facepalm issues

**Review by RipleyTom — APPROVED** [review]

**Comment by Kramerika** [comment]
I'm not sure how to get this to work with 3 microphones.  I am on Windows 10 with 3 Wii U USB microphones on RPCS3 Version: 0.0.27-f912aa7a Alpha | pull/13484.  They all show up as active microphones in Windows.  I have them configured in the RPCS3 audio settings using the standard OpenAL options for the first 3 microphones assigned to each one.  When I launch the game, two of them are active (separately) as mic 1 and mic 3, but one is not active for mic 2.  The even stranger thing is if I have only 2 mics connected then only one of them works but as mic 1 and mic 3 simultaneously.  Do I have something misconfigured?

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11084780/RPCS3.log)

**Comment by T3HM3T4L** [comment]
> I'm not sure how to get this to work with 3 microphones. I am on Windows 10 with 3 Wii U USB microphones on RPCS3 Version: 0.0.27-f912aa7a Alpha | pull/13484. They all show up as active microphones in Windows. I have them configured in the RPCS3 audio settings using the standard OpenAL options for the first 3 microphones assigned to each one. When I launch the game, two of them are active (separately) as mic 1 and mic 3, but one is not active for mic 2. The even stranger thing is if I have only 2 mics connected then only one of them works but as mic 1 and mic 3 simultaneously. Do I have something misconfigured?
> 
> [RPCS3.log](https://github.com/RPCS3/rpcs3/files/11084780/RPCS3.log)

It isn't you, this is the way multiple mics are handled in the main branch and for some reason that behaviour returned in this latest audio2 build it seems.

I confirmed it by testing the new build (exactly the same behaviour as you describe) versus the previous audio2 build I had (rpcs3-v0.0.27-14832-8180d634_linux64) where the 2 mics still work as intended.

File: rpcs3/Emu/Cell/Modules/cellMic.cpp [hunk 5]
```diff
@@ -134,48 +134,113 @@ void mic_context::operator()()
...
+			[[maybe_unused]] const u32 index = register_device(device_list[0]);
 		}
-		case microphone_handler::real_singstar:
-		case microphone_handler::rocksmith:
+	}
+}
+
+u32 mic_context::register_device(const std::string& device_name)
+{
+	usz index = 0;
```
Comment (RipleyTom) [inline rpcs3/Emu/Cell/Modules/cellMic.cpp:172]: should be initialized to mic_list.size()

**Comment by T3HM3T4L** [comment]
2 mic and 3 mic work again after this latest fix, still has performance hits on Linux.

[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11092838/RPCS3.log)

**Comment by Kramerika** [comment]
I tested the new build with 1, 2, and 3 microphones, and it worked fine with any combination.

**Comment by moronally** [comment]
Hey all - I believe I followed the instructions for updating rpcs3. I ended up at cirrus and tried to download the artifacts but I keep getting error messages. Are the files no longer available? Or does anyone have a backup available?

Thanks

**Pull request merged**